arroyo/backends/abstract.py

```diff
@@ -175,11 +187,69 @@ def member_id(self) -> str:
         raise NotImplementedError
 
 
+class ProducerFuture(Protocol, Generic[T]):
+    """
+    An abstract interface for a kind of Future. Stdlib futures are too slow to
+    construct, so we use these.
+    """
+
+    def done(self) -> bool:
+        ...
+
+    def result(self, timeout: float | None = None) -> T:
+        """
+        Return result or raise exception. May block, but does not have to.
+        """
+        ...
+
+    def set_result(self, result: T) -> None:
+        ...
+
+    def set_exception(self, exception: Exception) -> None:
+        ...
+
+
+class SimpleProducerFuture(Generic[T]):
+    """
+    A stub for concurrent.futures.Future that does not construct any Condition
+    variables, therefore is faster to construct.
+    """
+
+    def __init__(self) -> None:
+        self.result_value: T | None = None
+        self.result_exception: Exception | None = None
+
+    def done(self) -> bool:
+        return self.result_value is not None
```

Comment:
Should this also check the exception? E.g. is this future "done" if the exception is set?

Reply:
good catch